Coastal Adaptation

AdamTaylor9 opened the pull request.

AdamTaylor9 committed:
initial commit

AdamTaylor9 committed:
add method to properly assign hazard_threshold for direct_damage rule

AdamTaylor9 committed:
Merge branch 'development' into coastal_adaptation_integration

AdamTaylor9 committed:
change naming convention of flood protection to asset dictionary parquets

thomas-fred commented:
Hey Adam. I'm afraid I'd like to change the damage/adaptation files again, see here: https://github.com/nismod/jamaica-infrastructure/pull/88. I do think it's worth it though. I had a think about our conversation on supplying and using the flood threshold and decided that the neatest way was to write a new `hazard_damage_parameters.csv` file for each `direct_damage` rule invocation. This handles both the flood and cyclone cases and works for the nominal case and the perturbing/adaptation cases. I've also pulled out the parameters from the rule and the Python scripts into the config file so they can be varied by a user.

I'm just testing the PR at the moment, but will probably merge to development later today.

AdamTaylor9 committed:
Merge branch 'development' into coastal_adaptation_integration

AdamTaylor9 committed:
modify benefit cost ratio to process coastal adaptation case

AdamTaylor9 committed:
modify coastal adaptation bcr to only use threshold values for projected end year date

AdamTaylor9 committed:
create adapt_cost_npv columns for each rcp for coastal adaptation bcr

AdamTaylor9 committed:
created new rule to handle generation of combined coastal protection gpkg

AdamTaylor9 committed:
Merge branch 'development' into coastal_adaptation_integration

AdamTaylor9 committed:
modify count_coastal_assets script to produce one csv file instead with inclusion of asset costs

AdamTaylor9 committed:
added coastline protection length to coastline protection assets breakdown

AdamTaylor9 committed:
modified adaptation_options_costs to handle coastal adaptation case

AdamTaylor9 committed:
added asset cost to asset to coastal protection mappings

AdamTaylor9 committed:
fixed errors in direct damage calcs and asset to coastal protection mapping

AdamTaylor9 committed:
filtered out adaptation costs for assets not in the flood protection areas

AdamTaylor9 committed:
modified bcr estimation script  to process coastal adaptation case

AdamTaylor9 committed:
Merge branch 'development' into coastal_adaptation_integration

AdamTaylor9 committed:
modifed coastal adaptation feature generation to only generate for a set rcp, epoch & rp

AdamTaylor9 committed:
modified coastal adaptation rules to work with set rcp, epoch and rp

AdamTaylor9 committed:
fixed direct damages to work with new coastal protection results format

AdamTaylor9 committed:
check asset max cost column or min cost column if mean cost column not available

AdamTaylor9 committed:
modified benefit cost ratio to work with new coastal protection results format

AdamTaylor9 committed:
moved the multiplication of asset npvs from bcr script to adaptation options cost script

AdamTaylor9 committed:
bug fixes in bcr script

AdamTaylor9 committed:
Merge branch 'development' into coastal_adaptation_integration

AdamTaylor9 committed:
modified generate_coastal_adaptations.py to fix edge case errors

AdamTaylor9 committed:
add new snakemake rule to process coastal adaptation results for visualisation

AdamTaylor9 committed:
Merge branch 'development' into coastal_adaptation_integration

thomas-fred committed:
Add expand rule to run coastal adaptation for all layers

thomas-fred commented on workflow/5_adaptation/asset_to_coastal_protection_mapping.py:
```diff
@@ -0,0 +1,292 @@
+    
+    networks_csv = network_csv
+    networks = pd.read_csv(networks_csv)
+    networks = networks[networks["asset_description"] != "buildings"]
```
How come you filter out buildings here? Can we not run the process for them?

AdamTaylor9 replied:
I believe this was a remnant of when the script iterated through all the network layers and for my computers sake I skipped the buildings layer. The script should be able to process the buildings layer as well.

thomas-fred commented:
I've taken your input files and recreated the coastal protection zones with:
`snakemake -c1 results/coastal_protection_assets/Jamaica_coastal_protection_areas.gpkg` and that seems to work :+1:

thomas-fred commented:
I presume to run the workflow, I should just be triggering the BCR rule? i.e.
`snakemake -c1 results/adaptation_benefits_costs_bcr/flooding_waste_water_facilities_NWC_nodes_adaptation_costs_avoided_EAD_EAEL.csv`

If I try and run that, there are missing input files, specifically damage curve files for coastal flooding:
```
MissingInputException in rule direct_damage in file /home/fred/projects/jamaica/jamaica-infrastructure/workflow/1_damage/_damage.smk, line 131:
Missing input files for rule direct_damage:
    output: results/direct_damages/waste_water_facilities_NWC_nodes/waste_water_facilities_NWC_nodes_direct_damages_parameter_set_0.parquet
    wildcards: output_path=results, gpkg=waste_water_facilities_NWC, layer=nodes, parameter_set=parameter_set_0
    affected files:
        processed_data/damage_curves/damage_curves_water_coastal.xlsx
```

thomas-fred commented:
What's the purpose of the new damage curve files? For example, `damage_curves_energy_coastal.xlsx` appears to be the same as `damage_curves_energy_flooding.xlsx`. Can we reuse the old ones?

AdamTaylor9 commented:
> I presume to run the workflow, I should just be triggering the BCR rule? i.e. `snakemake -c1 results/adaptation_benefits_costs_bcr/flooding_waste_water_facilities_NWC_nodes_adaptation_costs_avoided_EAD_EAEL.csv`
> 
> If I try and run that, there are missing input files, specifically damage curve files for coastal flooding:
> 
> ```
> MissingInputException in rule direct_damage in file /home/fred/projects/jamaica/jamaica-infrastructure/workflow/1_damage/_damage.smk, line 131:
> Missing input files for rule direct_damage:
>     output: results/direct_damages/waste_water_facilities_NWC_nodes/waste_water_facilities_NWC_nodes_direct_damages_parameter_set_0.parquet
>     wildcards: output_path=results, gpkg=waste_water_facilities_NWC, layer=nodes, parameter_set=parameter_set_0
>     affected files:
>         processed_data/damage_curves/damage_curves_water_coastal.xlsx
> ```

You're right about running the BCR rule but, the hazard wildcard has to be 'coastal' instead of 'flooding', so the rule would resemble 

`snakemake -c1 results/adaptation_benefits_costs_bcr/coastal_waste_water_facilities_NWC_nodes_adaptation_costs_avoided_EAD_EAEL.csv`

AdamTaylor9 commented:
> What's the purpose of the new damage curve files? For example, `damage_curves_energy_coastal.xlsx` appears to be the same as `damage_curves_energy_flooding.xlsx`. Can we reuse the old ones?

It was because of the same me changing the hazard wildcard to 'coastal' as the rules were looking for damage curve files with 'coastal' instead of 'flooding' in the filename, so I figured the easiest way to resolve it was just make a file which matches the filename it was looking for.

thomas-fred committed:
Add explicit gdal dependency with recent version

Fixes missing proj DB error from splits script

thomas-fred committed:
Remove module level constant

thomas-fred committed:
Do not skip buildings

thomas-fred committed:
Add target rule for coastal protection BCR

thomas-fred committed:
Short circuit when no matching adaptation options

thomas-fred commented on workflow/5_adaptation/adaptation_options_costs.py:
```diff
@@ -110,10 +110,48 @@ def get_dimension_factor(x):
+    if asset_match.empty:
+        dimension = 1
+    else:
+        flood_id = int(asset_match[flood_id_col].iloc[0])
```
If I run this with `snakemake -c1 results/adaptation_costs/coastal_costs/airport_polygon_areas_adaptation_timeseries_and_npvs.csv` I get a `TypeError` as the input to the `int` function is `None` of `NoneType`. What should the behaviour in this case be?

thomas-fred replied:
```
  File "/data/jamaica/jamaica-infrastructure/workflow/5_adaptation/adaptation_options_costs.py", line 122, in get_coastal_dimension_factor
    flood_id = int(asset_match[flood_id_col].iloc[0])
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'NoneType'
```

thomas-fred replied:
Does this work as an alternative?

```
value = asset_match[flood_id_col].iloc[0]
if pd.isna(value):  # Handle NaN values explicitly
    dimension = 1
else:
    flood_id = int(value)
    ...
```

AdamTaylor9 replied:
Yes that should work fine, because if it can't find a flood id for that asset then the dimension should be set to 1.

thomas-fred committed:
Handle NoneType flood ID

thomas-fred committed:
Actually create flag file

thomas-fred committed:
Gracefully handle case of no adaptation options

thomas-fred committed:
Gracefully handle empty single failure scenario file

thomas-fred committed:
Add empty single failure scenario file (header only)

thomas-fred committed:
Add coastal adaptation example invocation

thomas-fred committed:
Tidy up -- remove unused imports, formatting

thomas-fred committed:
Reorder expressions for clarity

thomas-fred committed:
Improve readability of adaptation_options_costs function

Rename variables, reorder, reformat

thomas-fred committed:
Don't filter network layers by description

Instead, rely on asset_gpkg and asset_layer to filter network layers. If
we don't have any adaptation options for the assets in that layer, find
that out before short-circuiting.

thomas-fred committed:
Remove variable

thomas-fred committed:
Use original file name

thomas-fred committed:
Skip if (hazard) asset type lookup column is 'none'

thomas-fred committed:
Ignore sbatch jobfiles

thomas-fred committed:
Hardcode path for coastal protection visualisation rule

thomas-fred committed:
Use single line for shell command

Avoid issue with multi-line strings being truncated at line continuation

thomas-fred committed:
Refactor to reduce memory consumption

thomas-fred committed:
Fix logging error

thomas-fred committed:
Update shell blocks to use format rather than f-strings

There seems to be an issue with mixing Python variable interpolation ({DATA}),
literal strings ('areas') and snakemake template variables ({{input.script}})
within an f-string shell block. Using the format method on a string will
circumvent this.

thomas-fred committed:
Add a buildings specific preprocess rule

thomas-fred committed:
Fix line breaks

thomas-fred committed:
Format with black

thomas-fred committed:
Short circuit if no adaptation options available

thomas-fred committed:
Format with black

thomas-fred committed:
Factor out coastal protection to separate script

Find costs to assets avoided by construction of coastal defence. Save these
data in the file typically used for EAD and EAEL.

thomas-fred committed:
Add TODO

thomas-fred committed:
Use (avoided) damage files parameterised by epoch

thomas-fred committed:
Save more data to disk

- Include an `asset_type` field for the area layer
- Save the coastal defence area layer to both output files

thomas-fred committed:
Change edges -> areas

thomas-fred committed:
Split BCR into normal & coastal, factor out functions

thomas-fred committed:
Fix indent level error

thomas-fred committed:
Update docstrings

thomas-fred committed:
Constrain hazard wildcard

thomas-fred committed:
Use correct type (list[str]) for hazards

thomas-fred committed:
Fix typos

thomas-fred committed:
Find BCR for entire coastal protection zones

thomas-fred committed:
Find agg. avoided risk by coastal protection per RP

thomas-fred committed:
Extract coastal protection to own directory; tidy

thomas-fred committed:
Apply black

thomas-fred committed:
Rename ETL scripts

thomas-fred committed:
Remove redundant placeholder rules

thomas-fred committed:
Remove redundant changes to risk calculation

Previously, there were some changes to the `direct_damages` rule and associated
script which aimed to calculate direct damages given a coastal adaptation
option (e.g. sea wall). These are redundant, as downstream we assume all
protected assets have all of their risks averted -- the flood defences are
sized to exclude the maximum flood present in the coastal rasters.

thomas-fred committed:
Use correct argument name and path
